packages/react-router/lib/components.tsx

```diff
@@ -87,27 +87,36 @@ export function RouterProvider({
 
   let basename = router.basename || "/";
 
+  // The fragment and {null} here are important!  We need them to keep React 18's
+  // useId happy when we are server-rendering since we may have a <script> here
+  // containing the hydrated server-side staticContext (from StaticRouterProvider).
+  // useId relies on the component tree structure to generate deterministic id's
+  // so we need to ensure it remains remains the same on the client even though
+  // we don't need the <script> tag
   return (
-    <DataRouterContext.Provider
-      value={{
-        router,
-        navigator,
-        static: false,
-        // Do we need this?
-        basename,
-      }}
-    >
-      <DataRouterStateContext.Provider value={state}>
-        <Router
-          basename={router.basename}
-          location={router.state.location}
-          navigationType={router.state.historyAction}
-          navigator={navigator}
-        >
-          {router.state.initialized ? <Routes /> : fallbackElement}
-        </Router>
-      </DataRouterStateContext.Provider>
-    </DataRouterContext.Provider>
+    <>
+      <DataRouterContext.Provider
+        value={{
+          router,
+          navigator,
+          static: false,
+          // Do we need this?
+          basename,
+        }}
+      >
+        <DataRouterStateContext.Provider value={state}>
+          <Router
+            basename={router.basename}
+            location={router.state.location}
+            navigationType={router.state.historyAction}
+            navigator={navigator}
+          >
+            {router.state.initialized ? <Routes /> : fallbackElement}
+          </Router>
+        </DataRouterStateContext.Provider>
+      </DataRouterContext.Provider>
+      {null}
+    </>
```

Comment:
Ensure the structure generated by `RouterProvider` matches that generated by `StaticRouterProvider`

Reply:
I'm a little confused why we need the `null` since the [docs](https://beta.reactjs.org/reference/react/useId) indicate that only the parent path matters, and our `useId` calls would be down inside `<Routes/>`.  But without the null I constantly ran into hydration mismatches 😕 

<img width="802" alt="Screenshot 2023-01-03 at 4 23 01 PM" src="https://user-images.githubusercontent.com/1609022/210451906-b154eed4-4b23-4418-bd0a-b2c0d5fa03c4.png">

Reply:
Mind filing a question / bug over on the react repo?

Reply:
This seems expected - it might just be a docs issue: https://github.com/facebook/react/issues/22733.  Added my own comment in there suggesting a docs update